### ISSUE TITLE
Question Setting ef_search to different values does not affect number of results retrieved (Django 4.2 + PGVector + HNSW index)

### ISSUE THREAD

**Issue by rmincling** [pgvector/pgvector#675]
Hi all,

I have this simple Django model setup:

```
class AiVector(models.Model):
    # Embeddings
    embedding_sbert = VectorField(dimensions=768, null=True, blank=True) # using SBERT embedding (all-mpnet-base-v2)

    class Meta:
        indexes = [
            # SBERT
            HnswIndex(
                name='vector_sbert_index',
                fields=['embedding_sbert'],
                m=16,
                ef_construction=64, 
                opclasses=['vector_cosine_ops']  
            )]
```

I am trying to do a basic vector search (cosine) on the model with a simple question/embedding with the following code:
```
         with connection.cursor() as cursor:
                cursor.execute("SET LOCAL hnsw.ef_search = 40")
            self.matching_vectors = AiVector.objects \
                .annotate(distance=CosineDistance("embedding_sbert", text_query_embedding)) \
                .filter(distance__lte=0.8)
```
No matter what i set ef_search to, whether it be 1, 40, 60, 100, I always get the same number of results. What is the correct way to set ef_search?

Regards,

Rob

**Comment by jkatz** [comment]
@rmincling How many results are you getting?

(FWIW, it's been awhile since I've written heavy Django code, but I suspect that you're executing the query using two different cursors, which means it's unaware of the setting of `hnsw.ef_search` that you're explicitly setting.)

**Comment by rmincling** [comment]
I'm getting over 3000 results

**Comment by jkatz** [comment]
@rmincling In that case, this is because you haven't [set a `LIMIT`](https://docs.djangoproject.com/en/5.1/topics/db/queries/#limiting-querysets), and PostgreSQL is opting to use a table scan instead of an index scan. I suggest trying:

```python
            self.matching_vectors = AiVector.objects \
                .annotate(distance=CosineDistance("embedding_sbert", text_query_embedding)) \
                .filter(distance__lte=0.8)[:10]
```

to return 10 results.

**Comment by rmincling** [comment]
I will try that. How do i check if the hnsw index is being used at all?

**Comment by rmincling** [comment]
ok, so here is an interesting update, regardless of what LIMIT I used, the HNSW index was NOT being used (confirmed when using EXPLAIN ANALYZE. It was using a sequential scan.

So I tried this:
```
cursor.execute("SET LOCAL enable_seqscan TO off;")
```
And voila, the HNSW was used!

So now, when i change the value of ef_search the amount of results changes!
```
cursor.execute("SET LOCAL hnsw.ef_search = 100")
```

My next question is, I expect this vector table to have potentially 100k rows. So how do I decide if I should used HNSW index or not?

**Comment by jkatz** [comment]
> I expect this vector table to have potentially 100k rows. So how do I decide if I should used HNSW index or not?

In this case, PostgreSQL will likely opt to use the HNSW index for your query. You shouldn't need to set `enable_seqscan` (and for a production workload I strongly recommend against doing so).

**Comment by rmincling** [comment]
It seems the maximum value of ef_search is 1000, didn't notice that before. Wonder will that be increased in future versions of postgres/pgvector

**Comment by jkatz** [comment]
@rmincling If you're interested in discusing `hnsw.ef_search` limits, I'd recommend opening up a new issue with the stated use case as to why it'd be beneficial to have it increased. If you're now unblocked with your original question, I'd recommend closing out this issue. Thanks!

**Comment by rmincling** [comment]
Ok, will do!, Thanks for all of your help and insights
